sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala

```diff
@@ -0,0 +1,452 @@
+    tableIdent: TableIdentifier,
+    userSpecifiedSchema: Option[StructType],
+    provider: String,
+    options: Map[String, String],
+    ignoreIfExists: Boolean,
+    managedIfNoPath: Boolean)
+  extends RunnableCommand {
+
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    // Since we are saving metadata to metastore, we need to check if metastore supports
+    // the table name and database name we have for this query. MetaStoreUtils.validateName
+    // is the method used by Hive to check if a table name or a database name is valid for
+    // the metastore.
+    if (!CreateDataSourceTableUtils.validateName(tableIdent.table)) {
+      throw new AnalysisException(s"Table name ${tableIdent.table} is not a valid name for " +
+        s"metastore. Metastore only accepts table name containing characters, numbers and _.")
+    }
+    if (tableIdent.database.isDefined &&
+      !CreateDataSourceTableUtils.validateName(tableIdent.database.get)) {
+      throw new AnalysisException(s"Database name ${tableIdent.database.get} is not a valid name " +
+        s"for metastore. Metastore only accepts database name containing " +
+        s"characters, numbers and _.")
+    }
+
+    val tableName = tableIdent.unquotedString
+    val sessionState = sqlContext.sessionState
+
+    if (sessionState.catalog.tableExists(tableIdent)) {
+      if (ignoreIfExists) {
+        return Seq.empty[Row]
+      } else {
+        throw new AnalysisException(s"Table $tableName already exists.")
+      }
+    }
+
+    var isExternal = true
+    val optionsWithPath =
+      if (!options.contains("path") && managedIfNoPath) {
+        isExternal = false
+        options + ("path" -> sessionState.catalog.defaultTablePath(tableIdent))
+      } else {
+        options
+      }
+
+    // Create the relation to validate the arguments before writing the metadata to the metastore.
+    DataSource(
+      sqlContext = sqlContext,
+      userSpecifiedSchema = userSpecifiedSchema,
+      className = provider,
+      bucketSpec = None,
+      options = optionsWithPath).resolveRelation()
+
+    CreateDataSourceTableUtils.createDataSourceTable(
+      sqlContext = sqlContext,
+      tableIdent = tableIdent,
+      userSpecifiedSchema = userSpecifiedSchema,
+      partitionColumns = Array.empty[String],
+      bucketSpec = None,
+      provider = provider,
+      options = optionsWithPath,
+      isExternal = isExternal)
+
+    Seq.empty[Row]
+  }
+}
+
+/**
+ * A command used to create a data source table using the result of a query.
+ *
+ * Note: This is different from [[CreateTableAsSelect]]. Please check the syntax for difference.
+ * This is not intended for temporary tables.
+ *
+ * The syntax of using this command in SQL is:
+ * {{{
+ *   CREATE TABLE [IF NOT EXISTS] [db_name.]table_name
+ *   USING format OPTIONS ([option1_name "option1_value", option2_name "option2_value", ...])
+ *   AS SELECT ...
+ * }}}
+ */
+case class CreateDataSourceTableAsSelectCommand(
```

Comment:
as discussed offline, I think we should just merge CreateDataSourceTableAsSelectCommand and CreateDataSourceTableCommand